### PR TITLE
Fix #1747: Add default redirect binding and upgrade to use 2.3.0 Code…

### LINKED ISSUE #1747: docfx 1.18.3 crashes on Windows 10 with docfx-seed repo
I just upgraded to v2.18.3 (chocolatey package) and docfx now crashes when trying to run on the docfx-seed repository. Well, I guess it doesn't technically crash, but the logged error looks like an unhandled exception.

I'm running Windows 10 (creators) with Visual Studio 2017 Preview (15.3 preview 2).

```
[17-06-08 08:48:34.886]Info:Config file docfx.json found, start generating metadata...
[17-06-08 08:48:41.926]Info:Completed executing in 6865.4273 milliseconds.
[17-06-08 08:48:41.926]Error:System.AggregateException: One or more errors occurred. ---> Microsoft.DocAsCode.Exceptions.ExtractMetadataException: Error extracting metadata for {"Items":{"obj/api":["C:/Users/david/Documents/Junk/docfx-seed/src/SampleClass1/SampleClass1.sln"]},"ApiFolderName":"","TocFileName":"toc.yml","IndexFileName":".manifest","PreserveRawInlineComments":false,"ForceRebuild":false,"ShouldSkipMarkup":false,"UseCompatibilityFileName":false,"MSBuildProperties":{"Configuration":"Release"}}: Unable to generate spec reference for !:CounterSampleCalculator ---> Microsoft.DocAsCode.Exceptions.DocfxException: Unable to generate spec reference for !:CounterSampleCalculator ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSharpNameVisitor.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.NameVisitorCreator.GetName(ISymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSReferenceItemVisitor.AddLinkItems(INamedTypeSymbol symbol, Boolean withGenericeParameter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ReferenceItemVisitor.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.CompositeYamlModelGenerator.GenerateReferenceInternal(ISymbol symbol, ReferenceItem reference, SymbolVisitorAdapter adapter, Boolean asOverload)
   at Microsoft.DocAsCode.Metadata.ManagedReference.YamlModelGenerator.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters, Dictionary`2 references, SymbolVisitorAdapter adapter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters)
   --- End of inner exception stack trace ---
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitProperty(IPropertySymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitDescendants[T](IEnumerable`1 children, Func`2 getChildren, Func`2 filter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitNamespace(INamespaceSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitDescendants[T](IEnumerable`1 children, Func`2 getChildren, Func`2 filter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitAssembly(IAssemblySymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.GenerateYamlMetadata(Compilation compilation, IAssemblySymbol assembly, Boolean preserveRawInlineComments, String filterConfigFile, IReadOnlyDictionary`2 extensionMethods)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<GetMetadataFromProjectLevelCacheAsync>d__43`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<SaveAllMembersFromCacheAsync>d__33.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<ExtractMetadataAsync>d__20.MoveNext()
   --- End of inner exception stack trace ---
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<ExtractMetadataAsync>d__20.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Microsoft.DocAsCode.SubCommands.MetadataCommand.Exec(SubCommandRunningContext context)
   at Microsoft.DocAsCode.SubCommands.CompositeCommand.Exec(SubCommandRunningContext context)
   at Microsoft.DocAsCode.Program.ExecSubCommand(String[] args)
---> (Inner Exception #0) Microsoft.DocAsCode.Exceptions.ExtractMetadataException: Error extracting metadata for {"Items":{"obj/api":["C:/Users/david/Documents/Junk/docfx-seed/src/SampleClass1/SampleClass1.sln"]},"ApiFolderName":"","TocFileName":"toc.yml","IndexFileName":".manifest","PreserveRawInlineComments":false,"ForceRebuild":false,"ShouldSkipMarkup":false,"UseCompatibilityFileName":false,"MSBuildProperties":{"Configuration":"Release"}}: Unable to generate spec reference for !:CounterSampleCalculator ---> Microsoft.DocAsCode.Exceptions.DocfxException: Unable to generate spec reference for !:CounterSampleCalculator ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSharpNameVisitor.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.NameVisitorCreator.GetName(ISymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSReferenceItemVisitor.AddLinkItems(INamedTypeSymbol symbol, Boolean withGenericeParameter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ReferenceItemVisitor.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.CompositeYamlModelGenerator.GenerateReferenceInternal(ISymbol symbol, ReferenceItem reference, SymbolVisitorAdapter adapter, Boolean asOverload)
   at Microsoft.DocAsCode.Metadata.ManagedReference.YamlModelGenerator.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters, Dictionary`2 references, SymbolVisitorAdapter adapter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters)
   --- End of inner exception stack trace ---
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitProperty(IPropertySymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitDescendants[T](IEnumerable`1 children, Func`2 getChildren, Func`2 filter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitNamespace(INamespaceSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitDescendants[T](IEnumerable`1 children, Func`2 getChildren, Func`2 filter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitAssembly(IAssemblySymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.GenerateYamlMetadata(Compilation compilation, IAssemblySymbol assembly, Boolean preserveRawInlineComments, String filterConfigFile, IReadOnlyDictionary`2 extensionMethods)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<GetMetadataFromProjectLevelCacheAsync>d__43`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<SaveAllMembersFromCacheAsync>d__33.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<ExtractMetadataAsync>d__20.MoveNext()
   --- End of inner exception stack trace ---
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<ExtractMetadataAsync>d__20.MoveNext()<---



Build failed.
[17-06-08 08:48:41.941]Error:System.AggregateException: One or more errors occurred. ---> Microsoft.DocAsCode.Exceptions.ExtractMetadataException: Error extracting metadata for {"Items":{"obj/api":["C:/Users/david/Documents/Junk/docfx-seed/src/SampleClass1/SampleClass1.sln"]},"ApiFolderName":"","TocFileName":"toc.yml","IndexFileName":".manifest","PreserveRawInlineComments":false,"ForceRebuild":false,"ShouldSkipMarkup":false,"UseCompatibilityFileName":false,"MSBuildProperties":{"Configuration":"Release"}}: Unable to generate spec reference for !:CounterSampleCalculator ---> Microsoft.DocAsCode.Exceptions.DocfxException: Unable to generate spec reference for !:CounterSampleCalculator ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSharpNameVisitor.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.NameVisitorCreator.GetName(ISymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSReferenceItemVisitor.AddLinkItems(INamedTypeSymbol symbol, Boolean withGenericeParameter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ReferenceItemVisitor.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.CompositeYamlModelGenerator.GenerateReferenceInternal(ISymbol symbol, ReferenceItem reference, SymbolVisitorAdapter adapter, Boolean asOverload)
   at Microsoft.DocAsCode.Metadata.ManagedReference.YamlModelGenerator.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters, Dictionary`2 references, SymbolVisitorAdapter adapter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters)
   --- End of inner exception stack trace ---
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitProperty(IPropertySymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitDescendants[T](IEnumerable`1 children, Func`2 getChildren, Func`2 filter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitNamespace(INamespaceSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitDescendants[T](IEnumerable`1 children, Func`2 getChildren, Func`2 filter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitAssembly(IAssemblySymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.GenerateYamlMetadata(Compilation compilation, IAssemblySymbol assembly, Boolean preserveRawInlineComments, String filterConfigFile, IReadOnlyDictionary`2 extensionMethods)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<GetMetadataFromProjectLevelCacheAsync>d__43`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<SaveAllMembersFromCacheAsync>d__33.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<ExtractMetadataAsync>d__20.MoveNext()
   --- End of inner exception stack trace ---
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<ExtractMetadataAsync>d__20.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Microsoft.DocAsCode.SubCommands.MetadataCommand.Exec(SubCommandRunningContext context)
   at Microsoft.DocAsCode.SubCommands.CompositeCommand.Exec(SubCommandRunningContext context)
   at Microsoft.DocAsCode.Program.ExecSubCommand(String[] args)
---> (Inner Exception #0) Microsoft.DocAsCode.Exceptions.ExtractMetadataException: Error extracting metadata for {"Items":{"obj/api":["C:/Users/david/Documents/Junk/docfx-seed/src/SampleClass1/SampleClass1.sln"]},"ApiFolderName":"","TocFileName":"toc.yml","IndexFileName":".manifest","PreserveRawInlineComments":false,"ForceRebuild":false,"ShouldSkipMarkup":false,"UseCompatibilityFileName":false,"MSBuildProperties":{"Configuration":"Release"}}: Unable to generate spec reference for !:CounterSampleCalculator ---> Microsoft.DocAsCode.Exceptions.DocfxException: Unable to generate spec reference for !:CounterSampleCalculator ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSharpNameVisitor.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.NameVisitorCreator.GetName(ISymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSReferenceItemVisitor.AddLinkItems(INamedTypeSymbol symbol, Boolean withGenericeParameter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ReferenceItemVisitor.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.CompositeYamlModelGenerator.GenerateReferenceInternal(ISymbol symbol, ReferenceItem reference, SymbolVisitorAdapter adapter, Boolean asOverload)
   at Microsoft.DocAsCode.Metadata.ManagedReference.YamlModelGenerator.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters, Dictionary`2 references, SymbolVisitorAdapter adapter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters)
   --- End of inner exception stack trace ---
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.AddSpecReference(ISymbol symbol, IReadOnlyList`1 typeGenericParameters, IReadOnlyList`1 methodGenericParameters)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitProperty(IPropertySymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitNamedType(INamedTypeSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitDescendants[T](IEnumerable`1 children, Func`2 getChildren, Func`2 filter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitNamespace(INamespaceSymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitDescendants[T](IEnumerable`1 children, Func`2 getChildren, Func`2 filter)
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitAssembly(IAssemblySymbol symbol)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.GenerateYamlMetadata(Compilation compilation, IAssemblySymbol assembly, Boolean preserveRawInlineComments, String filterConfigFile, IReadOnlyDictionary`2 extensionMethods)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<GetMetadataFromProjectLevelCacheAsync>d__43`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<SaveAllMembersFromCacheAsync>d__33.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<ExtractMetadataAsync>d__20.MoveNext()
   --- End of inner exception stack trace ---
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<ExtractMetadataAsync>d__20.MoveNext()<---

	0 Warning(s)
	1 Error(s)
```

### PR DESCRIPTION
**Pull request by vicancy**
…Analysis for better logging

https://github.com/dotnet/roslyn/issues/19978#issuecomment-306892784
https://github.com/dotnet/roslyn/issues/2779

### CODE REVIEW

File: src/docfx/app.config [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<configuration>
+  <runtime>
+    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
+      <dependentAssembly>
+        <assemblyIdentity name="Microsoft.Build" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
+        <bindingRedirect oldVersion="14.0.0.0" newVersion="15.1.0.0" />
```
Comment (superyyrrzz) [inline src/docfx/app.config]: use range for oldVersion?

**Comment by hellosnow** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/Microsoft.DocAsCode.Metadata.ManagedReference.csproj [hunk 2]
```diff
@@ -14,15 +14,14 @@
...
-    <PackageReference Include="Microsoft.CodeAnalysis" Version="2.2.0" />
-    <PackageReference Include="Microsoft.CodeAnalysis.Common" Version="2.2.0" />
-    <PackageReference Include="Microsoft.CodeAnalysis.VisualBasic" Version="2.2.0" />
-    <PackageReference Include="Microsoft.CodeAnalysis.VisualBasic.Workspaces" Version="2.2.0" />
-    <PackageReference Include="Microsoft.CodeAnalysis.VisualBasic.Features" Version="2.2.0" />
-    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="2.2.0" />
-    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="2.2.0" />
-    <PackageReference Include="Microsoft.CodeAnalysis.Workspaces.Common" Version="2.2.0" />
-    <PackageReference Include="Microsoft.Net.Compilers" Version="2.2.0" />
+    <PackageReference Include="Microsoft.CodeAnalysis" Version="2.3.0-beta1" />
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/Microsoft.DocAsCode.Metadata.ManagedReference.csproj]: >beta1 [](start = 70, length = 5)

Question, why don't we wait for a stable release?

Comment (vicancy) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/Microsoft.DocAsCode.Metadata.ManagedReference.csproj]: 2.3 contains much more useful information for debugging https://github.com/dotnet/roslyn/issues/19978#issuecomment-306892784

Comment (vicancy) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/Microsoft.DocAsCode.Metadata.ManagedReference.csproj]: Yes it does look strange to use a prerelease version in release version of docfx... use 2.2.0 instead

**Pull request merged**